File: backend/src/main/java/com/transcendence/common/search/FilterOperator.java
package com.transcendence.common.search;

public enum FilterOperator {
    EQ,          // coluna = :param
    NEQ,         // coluna <> :param
    GT,          // coluna > :param
    GTE,         // coluna >= :param
    LT,          // coluna < :param
    LTE,         // coluna <= :param
    BEFORE,      // coluna < :param      (alias semântico de LT para datas)
    AFTER,       // coluna > :param      (alias semântico de GT para datas)
    BETWEEN,     // coluna BETWEEN :param1 AND :param2        — ranges (dates, numbers)
    CONTAINS,    // coluna ILIKE :param  (com % wrapping)
    STARTS_WITH, // coluna ILIKE :param%                      — prefix search
    ENDS_WITH,   // coluna ILIKE %:param                      — suffix search
    IS_NULL,     // coluna IS NULL       (sem parâmetro)
    IS_NOT_NULL, // coluna IS NOT NULL   (sem parâmetro)
    IN,          // coluna = ANY(:param) (para listas)
    NOT_IN       // coluna <> ALL(:param) or NOT IN (:param)  — complement of IN
}

File: backend/src/Common/Services/Search/Model/FilterOperator.cs
namespace Trippie.Common.Services.Search.Model;

/// <summary>
/// All the comparison operators a client can use in a filter.
/// </summary>
/// <remarks>
/// These map 1:1 to the Java enum. Names use PascalCase here (C# convention)
/// instead of SCREAMING_SNAKE_CASE (Java convention).
///
/// The compiler turns each value into a SQL fragment. For example:
///   <list type="bullet">
///     <item><see cref="Equal"/>          -> "col = @p"</item>
///     <item><see cref="Contains"/>       -> "col ILIKE @p ESCAPE '\'"</item>
///     <item><see cref="In"/>             -> "col = ANY(@p)"  (Postgres)</item>
///   </list>
/// </remarks>
public enum FilterOperator
{
    // ----- standard comparisons -----
    Equal,              // col = @p
    NotEqual,           // col <> @p
    GreaterThan,        // col > @p
    GreaterThanOrEqual, // col >= @p
    LessThan,           // col < @p
    LessThanOrEqual,    // col <= @p

    // ----- semantic aliases for dates (same SQL as GT/LT, clearer in payloads) -----
    After,
    Before,

    // ----- range (requires BOTH 'value' and 'valueTo') -----
    Between, // col BETWEEN @p1 AND @p2

    // ----- string pattern matching (uses ILIKE on Postgres) -----
    Contains,   // %value%
    StartsWith, // value%
    EndsWith,   // %value

    // ----- null checks (no parameter is bound for these) -----
    IsNull,
    IsNotNull,

    // ----- list operators -----
    In,    // col = ANY(@p)
    NotIn  // col <> ALL(@p)
}
